sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -129,6 +128,14 @@ trait CheckAnalysis {
             failAnalysis(
               s"unresolved operator ${operator.simpleString}")
 
+          case o if o.expressions.exists(!_.deterministic) &&
+            !o.isInstanceOf[Project] && !o.isInstanceOf[Filter] =>
+            failAnalysis(
```

Comment:
Hi, @cloud-fan .Can it support for join operation? Sometimes we can use some `non deterministic ` expression to eval some pointless join keys(with respect to business logic) avoiding data skew.
For example
```sql
SELECT src.key, src.value, src1.value 
FROM src 
JOIN src1
ON UPPER((CASE WHEN (src.key IS NULL OR src.key = '' ) THEN CAST( (-RAND() * 10000000 ) AS string ) ELSE src.key END )) = UPPER(src1.key)
```
What do you think?